Add Rusanov boundary correction for the relativistic Euler system

fmahebert opened the pull request:
## Proposed changes

Depends on PRS #2967 and #2969 

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

fmahebert committed:
Add Rusanov boundary correction for RelativisticEuler

fmahebert committed:
Add Python init files

fmahebert committed:
Add temporary return args to RelEuler TimeDerivativeTerms

Return the lapse, shift, and spatial metric from ComputeTimeDerivative's
temporary args mechanism. This will enable computing the system's
characteristic speeds on the boundary.

fmahebert committed:
Adjust metric in bdry correction helper to be physically reasonable

nilsdeppe commented on tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/BoundaryCorrections/Test_Rusanov.cpp:
```diff
@@ -0,0 +1,154 @@
+      helpers::Tags::Range<hydro::Tags::SpecificInternalEnergy<DataVector>>,
+      helpers::Tags::Range<hydro::Tags::SpecificEnthalpy<DataVector>>,
+      helpers::Tags::Range<hydro::Tags::SpatialVelocity<DataVector, Dim>>>
+      ranges{std::array<double, 2>{{1e-6, 1.0}},
```
I'm not sure what the lower range should be. `1.0e-30` still seems reasonable to test.

Also `1e-6` -> `1.0e-6`

nilsdeppe reviewed (commented):
Overall looks good, one small nitpick. Feel free to squash immediately

fmahebert commented:
Rebased on develop now that no longer dependent.

I've also squashed in a few small changes:
- changed bounds/notation as per Nils's review
- cleaned up computation of max char speeds in the python comparison implementation

nilsvu commented on tests/Unit/Helpers/Evolution/DiscontinuousGalerkin/BoundaryCorrections.hpp:
```diff
@@ -290,6 +291,27 @@ void test_boundary_correction_conservation_impl(
         get<inv_spatial_metric>(exterior_fields_on_face));
+
+    // if dg_package_data depends on the (not inverted) spatial metric, we
+    // also adjust this. Note we that for testing purposes, the metric and
```
<s>we</s>

nilsvu commented on src/Evolution/Systems/RelativisticEuler/Valencia/BoundaryCorrections/Rusanov.cpp:
```diff
@@ -0,0 +1,250 @@
+        equation_of_state) const noexcept {
+  {
+    // First compute v^2 and cs^2
+    Scalar<DataVector>& spatial_velocity_squared =
```
(optional) `auto&`, and would be helpful to add a comment that the two return variables are used as memory buffer

nilsvu commented on tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/BoundaryCorrections/Rusanov.py:
```diff
@@ -0,0 +1,149 @@
+import Evolution.Systems.RelativisticEuler.Valencia.TestFunctions as valencia
+
+
+def dg_package_data_tilde_d(
+    tilde_d, tilde_tau, tilde_s, flux_tilde_d, flux_tilde_tau, flux_tilde_s,
+    lapse, shift, spatial_metric, rest_mass_density, specific_internal_energy,
+    specific_enthalpy, spatial_velocity, normal_covector, normal_vector,
+    mesh_velocity, normal_dot_mesh_velocity, use_polytropic_fluid):
```
(optional) you can abbreviate all of these unused args with `dg_package_data_tilde_d(tilde_d, *args, **kwargs):`

fmahebert replied:
This is new info to me, so thanks for the pointer! I think I will leave as-is for consistency with elsewhere, though?

nilsvu replied:
Up to you, but I don't think we have to prioritize consistency here. Also this is already used in `FishboneMoncriefDisk.py`, `KomissarovShock.py` and `FirstOrderInternalPenalty.py` (disclaimer: I wrote 2/3 of those).

nilsvu commented on tests/Unit/Evolution/Systems/RelativisticEuler/Valencia/BoundaryCorrections/Rusanov.py:
```diff
@@ -0,0 +1,149 @@
+        return -0.5 * (interior_normal_dot_flux_tilde_d +
+                       exterior_normal_dot_flux_tilde_d) - 0.5 * np.maximum(
+                           interior_abs_char_speed, exterior_abs_char_speed
+                       ) * (exterior_tilde_d - interior_tilde_d)
```
(optional) math often formats nicer in Python when you wrap it in parens: `return (-0.5 * ...)`. Could help with a bunch of expressions in this file.

fmahebert replied:
Doing this here gave a small but appreciable formatting improvement, so I went with it.

fmahebert commented:
Fixups posted

nilsvu reviewed (commented):
I'm happy with squashing the fixups. Also feel free to squash if you decide to go with the `*args` syntax.

nilsdeppe commented:
fixups look fine. I personally have started avoiding `auto` more because having the type can help with readability and also helps with auto completion in code editors, but I'm fine either way here.

fmahebert commented:
Okay, since @nilsdeppe agrees with me on preferring explicit types, and @nilsleiffischer's request was optional, I reverted back to writing `Scalar<DataVector>` on the two temporaries.

Squashed the fixups and rebased on develop.

nilsvu reviewed (approved):
Re `auto`: We should probably follow https://google.github.io/styleguide/cppguide.html#Type_deduction more closely :) (use `auto` only where it helps readability)

nilsvu commented:
(Note that our current guidelines contradict the Google style guide, so we may want to change them https://spectre-code.org/code_review_guide.html)
